flutter/lib/resources/resource_manager.dart

```diff
@@ -71,26 +86,42 @@ class ResourceManager {
         await Directory(path).exists();
   }
 
-  void handleResources(List<String> resources, bool purgeOldCache) async {
+  Future<bool> isChecksumMatched(String filePath, String md5Checksum) async {
+    var fileStream = File(filePath).openRead();
+    final checksum = (await md5.bind(fileStream).first).toString();
+    return checksum == md5Checksum;
+  }
+
+  void handleResources(List<Resource> resources, bool purgeOldCache) async {
     _progressString = '0%';
     _done = false;
     _onUpdate();
 
-    var internetResources = <String>[];
+    var internetResources = <Resource>[];
     for (final resource in resources) {
-      if (resource.startsWith(_applicationDirectoryPrefix)) continue;
-      if (isInternetResource(resource)) {
+      if (resource.path.startsWith(_applicationDirectoryPrefix)) continue;
+      if (isInternetResource(resource.path)) {
         internetResources.add(resource);
         continue;
       }
-      throw 'forbidden path: $resource (only http://, https:// and app:// resources are allowed)';
+      throw 'forbidden path: ${resource.path} (only http://, https:// and app:// resources are allowed)';
     }
 
-    await cacheManager.cache(internetResources, (double val) {
+    final internetPaths = internetResources.map((e) => e.path).toList();
+    await cacheManager.cache(internetPaths, (double val) {
       _progressString = '${(val * 100).round()}%';
       _onUpdate();
     }, purgeOldCache);
 
+    final checksumFailed = await validateResourcesChecksum(resources);
+    if (checksumFailed.isNotEmpty) {
+      final mismatchedPaths = checksumFailed.map((e) => '\n${e.path}').join();
+      throw 'Checksum validation failed for: $mismatchedPaths';
```

Comment:
I moved the checksum validation into the `handleResources()` method. It will throw an exception if checksum validation failed. I am expected it will show the `ExceptionWidget` when the exception is thrown but the app just crashes. The same for the case `forbidden path` exception some lines above.

@d-uzlov Do you have an idea why the app does not show the exception message in the UI, and how can we fix it?

Reply:
`resourceManager.handleResources` is an async function without `Future` object.
We don't `await` it, so the exception that is thrown inside of it is received by the Flutter framework itself, and it doesn't know how to handle it.

Reply:
I think we should be able to wrap `handleResources` in a try-catch block (maybe create a wrapping function) that would create error dialogues when an exception is thrown from `handleResources` or from any of the functions that it calls.

I think that's actually pretty major issue because there are other things that rely of exceptions. It's bad that the app could crash without an error message when anything goes wrong in the `handleResources` function.

Reply:
Currently when `ResourceManager` does anything it updates `BenchmarkState` and calls `_onUpdate()`.
We can probably keep it this way, and add some additional field, like `pendingException`.

Reply:
I'm not sure how hard it would be to fix it exactly. Maybe we should create a separate issue for this.

Reply:
It seems that handle the exception in `async` function will takes time so I created new issue for it https://github.com/mlcommons/mobile_app_open/issues/307.